cns/service/main.go

```diff
@@ -773,18 +779,81 @@ func InitializeMultiTenantController(ctx context.Context, httpRestService cns.HT
 	return nil
 }
 
-// initializeCRD state
-func InitializeCRDState(ctx context.Context, httpRestService cns.HTTPService, cnsconfig configuration.CNSConfig) error {
-	var requestController singletenantcontroller.RequestController
+type nodeNetworkConfigGetter interface {
+	Get(context.Context) (*v1alpha.NodeNetworkConfig, error)
+}
 
-	logger.Printf("[Azure CNS] Starting request controller")
+type ncStateReconciler interface {
+	ReconcileNCState(ncRequest *cns.CreateNetworkContainerRequest, podInfoByIP map[string]cns.PodInfo, scalar v1alpha.Scaler, spec v1alpha.NodeNetworkConfigSpec) error
+}
 
-	kubeConfig, err := kubecontroller.GetKubeConfig()
+// TODO(rbtr) where should this live??
+// InitCNS initializes cns by passing pods and a createnetworkcontainerrequest
+func initCNS(ctx context.Context, cli nodeNetworkConfigGetter, ncReconciler ncStateReconciler) error {
+	// Get nnc using direct client
+	nnc, err := cli.Get(ctx)
 	if err != nil {
-		logger.Errorf("[Azure CNS] Failed to get kubeconfig for request controller: %v", err)
+		// If the CRD is not defined, exit
+		if crd.IsNotDefined(err) {
+			logger.Errorf("CRD is not defined on cluster: %v", err)
+			os.Exit(1)
+		}
+
+		if nnc == nil {
+			logger.Errorf("NodeNetworkConfig is not present on cluster")
+			return nil
+		}
+
+		// If instance of crd is not found, pass nil to CNSClient
+		if client.IgnoreNotFound(err) == nil {
+			err = ncReconciler.ReconcileNCState(nil, nil, nnc.Status.Scaler, nnc.Spec)
+			if err != nil {
+				return errors.Wrap(err, "failed to reconcile NC state")
+			}
+			return nil
```

Comment:
Any reason not to just 
```go
return ncReconciler.ReconcileNCState(nil, nil, nnc.Status.Scaler, nnc.Spec)
```

Otherwise, if you must wrap the error, `errors.Wrap` returns nil if `err` is nil. You could just
```go
err = ncReconciler.ReconcileNCState(nil, nil, nnc.Status.Scaler, nnc.Spec)
return errors.Wrap(err, "failed to reconcile NC state")
```

Reply:
> errors.Wrap returns nil if err is nil
I have (ab)used this in other places, wasn't sure how others felt about it though. I prefer it and will update this